### PR TITLE
Change caching middleware, add Redis as storage

### PR DESCRIPTION
**Pull request by EnoRage**
Now Redis is used as storage for caching middleware
Main reason: now we can add multiple instances of API with the same cache

### CODE REVIEW

File: pkg/response_cache/cache.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+type cachedWriter struct {
+	gin.ResponseWriter
+	status  int
+	written bool
+	expire  int64
+	key     string
+}
+
+var _ gin.ResponseWriter = &cachedWriter{}
```
Comment (prazd) [inline pkg/response_cache/cache.go]: What is this for?

File: storage/storage.go [hunk 2]
```diff
@@ -5,7 +5,8 @@ import (
 )
 
 type SaveResult string
-const(
+
```
Comment (prazd) [inline storage/storage.go]: nice!

File: storage/storage.go [hunk 3]
```diff
@@ -28,3 +30,9 @@ type Market interface {
 	SaveRates(rates watchmarket.Rates, pl ProviderList) map[SaveResult]int
 	GetRate(currency string) (*watchmarket.Rate, error)
 }
+
+type Middleware interface {
```
Comment (arein) [inline storage/storage.go]: nit: i think `Middleware` for this interface is misleading.

File: storage/middleware.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	CacheResponse struct {
+		Status int
+		Header http.Header
+		Data   []byte
+	}
+
+	CacheData struct {
+		Data    CacheResponse
+		Expired int64
+	}
```
Comment (arein) [inline storage/middleware.go]: we have multiple types of cache in this application and even this package. So `CacheData` is a bit overloaded here.

File: api/middleware/cache.go [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+		b, _ = ioutil.ReadAll(c.Request.Body)
+		// Restore the io.ReadCloser to its original state
+		c.Request.Body = ioutil.NopCloser(bytes.NewBuffer(b))
+	}
+	hash := sha1.Sum(append([]byte(url), b...))
+	return base64.URLEncoding.EncodeToString(hash[:])
+}
+
+// CacheMiddleware encapsulates a gin handler function and caches the response with an expiration time.
+func CacheMiddleware(expiration int64, handle gin.HandlerFunc) gin.HandlerFunc {
```
Comment (arein) [inline api/middleware/cache.go]: nit: `expirationSeconds` or something with the unit would be better. Also `expiration` is the wrong name here. You are using it as a `maxAge`. So I'd suggest `maxAgeSeconds`

File: api/middleware/cache.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+func generateKey(c *gin.Context) string {
+	url := c.Request.URL.String()
+	var b []byte
+	if c.Request.Body != nil {
+		b, _ = ioutil.ReadAll(c.Request.Body)
+		// Restore the io.ReadCloser to its original state
+		c.Request.Body = ioutil.NopCloser(bytes.NewBuffer(b))
+	}
```
Comment (arein) [inline api/middleware/cache.go]: do you need to strip whitespace here?

File: api/middleware/cache.go [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+	hash := sha1.Sum(append([]byte(url), b...))
+	return base64.URLEncoding.EncodeToString(hash[:])
+}
+
+// CacheMiddleware encapsulates a gin handler function and caches the response with an expiration time.
```
Comment (arein) [inline api/middleware/cache.go]: this is HTTP/Gin specific. Would be better to reflect that in the name IMO e.g. `GinCachingMiddleware` or `HttpCachingMiddleware` to distinguish from other caching layers used here.

File: api/marketdata.go [hunk 8]
```diff
@@ -37,12 +36,11 @@ type Coin struct {
 }
 
 func SetupMarketAPI(router gin.IRouter, db storage.Market, charts *market.Charts, ac assets.AssetClient) {
-	router.Use(ginutils.TokenAuthMiddleware(viper.GetString("market.auth")))
 	// Ticker
 	router.POST("/ticker", getTickersHandler(db))
 	// Charts
-	router.GET("/charts", gincache.CacheMiddleware(time.Minute*10, getChartsHandler(charts)))
-	router.GET("/info", gincache.CacheMiddleware(time.Minute*5, getCoinInfoHandler(charts, ac)))
+	router.GET("/charts", middleware.CacheMiddleware(minute*5, getChartsHandler(charts)))
```
Comment (arein) [inline api/marketdata.go]: just curious: how often is the `time_start` parameter here the same roughly aka what do you roughly expect the cache hit rate to be?

File: api/middleware/cache.go [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+func CacheMiddleware(expiration int64, handle gin.HandlerFunc) gin.HandlerFunc {
+	if cache == nil {
+		logger.Fatal("gin cache middleware is created with empty cache")
+	}
+	return func(c *gin.Context) {
+		defer c.Next()
+		key := generateKey(c)
+		mc, err := cache.Get(key)
+		if err != nil || mc.Data.Data == nil || time.Now().Unix()-mc.Expired > expiration {
```
Comment (arein) [inline api/middleware/cache.go]: `mc.Data.Data` this hints at the naming being poor. Should be possible to find more descriptive names than a nested `Data`

Comment (arein) [reply api/middleware/cache.go]: Consider moving `time.Now().Unix()-mc.Expired > expiration` into a `isExpired` variable so your reader doesn't have to do math

File: api/middleware/cache.go [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+			writer := NewCachedWriter(expiration, c.Writer, key)
+			c.Writer = writer
+			handle(c)
+
+			if c.IsAborted() {
+				cache.Delete(key)
+			}
```
Comment (arein) [inline api/middleware/cache.go]: why?

File: api/middleware/cache.go [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+		c.Writer.WriteHeader(mc.Data.Status)
+		for k, vals := range mc.Data.Header {
+			for _, v := range vals {
+				c.Writer.Header().Set(k, v)
+			}
+		}
+		_, err = c.Writer.Write(mc.Data.Data)
+		if err != nil {
+			cache.Delete(key)
+			logger.Error(err, "cannot write data", mc)
```
Comment (arein) [inline api/middleware/cache.go]: can you explain this logic?

File: api/middleware/cache.go (hunk 10 above)
Comment (EnoRage) [reply api/middleware/cache.go]: if context is aborted during first response - we will delete cache

File: api/middleware/cache.go (hunk 6 above)
Comment (EnoRage) [reply api/middleware/cache.go]: what do you mean?

File: api/middleware/cache.go (hunk 11 above)
Comment (EnoRage) [reply api/middleware/cache.go]: Basically we cache the whole response with headers

File: storage/middleware.go (hunk 4 above)
Comment (EnoRage) [reply storage/middleware.go]: Can you suggest something more suitable, please?

File: storage/storage.go (hunk 3 above)
Comment (EnoRage) [reply storage/storage.go]: Can you suggest something more suitable, please?

File: api/marketdata.go (hunk 8 above)
Comment (EnoRage) [reply api/marketdata.go]: Yep, approximately

File: api/middleware/cache.go (hunk 7 above)
Comment (EnoRage) [reply api/middleware/cache.go]: Thank you! Good suggestion, will be resolved soon

File: api/middleware/cache.go (hunk 5 above)
Comment (EnoRage) [reply api/middleware/cache.go]: Thank you! Good suggestion, will be resolved soon

File: api/middleware/cache.go (hunk 9 above)
Comment (EnoRage) [reply api/middleware/cache.go]: Thank you! Good suggestion, will be resolved soon

**Comment by EnoRage** [comment]
> curious why you decided against putting the actual data in the cache?
> 
> Especially with the `/charts` route and its _time_start_ parameter it seems that the cache hit rate will be pretty low here.
> 
> Curious: how many hosts/containers does the API currently run on?

It was implemented not only for charts, there are 3 containers at API usually

**Comment by EnoRage** [comment]
You are also right about charts caching. I suggest that the caching must be implemented on the provider level

**Pull request closed without merging**